modnet/fit_genetic.py

```diff
@@ -0,0 +1,246 @@
+        child: List,
+        prob_mut: float = 0.8
+        )->None:
+
+        """Performs mutation in the genetic information in order to maintain diversity in the population. 
+        
+        Paramters:
+            child: List containing the genetic information of the 'child'.
+            prob_mut: Probability the mutation occurs.
+        """
+
+        for c in range(0, len(child)):
+            if np.random.rand() > prob_mut:
+                if child[c][0] < int(0.9*len(self.X_train.get_optimal_descriptors())):
+                    child[c][0] = int(child[c][0] + 10*randint(1, 6))
+                else:
+                    child[c][0] = int(child[c][0] - 10*randint(1, 6))
+        return child
+
+
+    def function_fitness(
+        self,
+        pop: List,
+        X_train: MODData,
+        y_train: pd.DataFrame,
+        X_val: MODData,
+        y_val: pd.DataFrame
+        )->None:
+
+        """Calculates the fitness of each model, which has the parameters contained in the pop argument. The function returns a list containing respectively the MSE calculated on the validation set, the model, and the parameters of that model.
+        
+        Parameters:
+            pop: List containing the genetic information (i.e., the parameters) of the model.
+            X_train: Input data of the training set.
+            y_train: Target values of the training set.
+            X_val: Input data of the validation set.
+            y_val: Target values of the validation set.
+        """
+
+        self.fitness = []
+        j = 0
+        rlr = keras.callbacks.ReduceLROnPlateau(
+            monitor="loss",
+            factor=0.5,
+            patience=20,
+            verbose=0,
+            mode="auto",
+            min_delta=0,
+        )
+        es = keras.callbacks.EarlyStopping(
+            monitor="loss",
+            min_delta=0.001,
+            patience=300,
+            verbose=0,
+            mode="auto",
+            baseline=None,
+            restore_best_weights=True,
+        )
+        callbacks = [rlr, es]
+        for w in self.pop:
+            modnet_model = MODNetModel([[['BV_Ea']]], {'BV_Ea':1}, n_feat=w[0], num_neurons=[[int(w[1])],[int(w[1]*w[2])],[int(w[1]*w[2]*w[3])],[int(w[1]*w[2]*w[3]*w[4])]], act=w[5])
+            try:
+                modnet_model.fit(X_train,val_fraction=0, val_key='BV_Ea',loss=w[6], lr=w[8], epochs = 250, batch_size = w[9], xscale=w[7], callbacks=callbacks, verbose=0)
+                modnet_model.fit(X_train,val_fraction=0, val_key='BV_Ea',loss=w[6], lr=w[8], epochs = 250, batch_size = 2*w[9], xscale=w[7], callbacks=callbacks, verbose=0)
+                modnet_model.fit(X_train,val_fraction=0, val_key='BV_Ea',loss=w[6], lr=w[8], epochs = 250, batch_size = 4*w[9], xscale=w[7], callbacks=callbacks, verbose=0)
+                modnet_model.fit(X_train,val_fraction=0, val_key='BV_Ea',loss=w[6], lr=w[8], epochs = 250, batch_size = 8*w[9], xscale=w[7], callbacks=callbacks, verbose=0)
```

Comment:
How robust is the batch-size increasing to small datasets? If I have 1000 samples and the batch size goes above 1000, does tensorflow crash or will it max out at 1000?

```suggestion
        for i in range(4):
                modnet_model.fit(X_train,val_fraction=0, val_key='BV_Ea',loss=w[6], lr=w[8], epochs = 250, batch_size = (2**i)*w[9], xscale=w[7], callbacks=callbacks, verbose=0)
```

Also the `val_key` here looks hard coded to your dataset, this will need to be fixed

Reply:
I wonder if we should actually handle this batch size increase as a learning schedule inside `MODNetModel.fit(...)` directly? This would then allow us to save Adam weights between resets